clash-lib/src/Clash/Primitives/Xilinx/ClockGen.hs

```diff
@@ -0,0 +1,126 @@
+{-|
+  Copyright   :  (C) 2023, QBayLogic B.V.
+                     2023, Google Inc.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox template functions for
+  Clash.Xilinx.ClockGen.{clockWizard,clockWizardDifferential}
+-}
+
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Primitives.Xilinx.ClockGen where
+
+import Clash.Backend                      (Backend)
+import Clash.Netlist.BlackBox.Util        (exprToString)
+import Clash.Netlist.Types
+import Clash.Netlist.Util                 (stripVoid)
+
+import Control.Monad.State                (State)
+import qualified Data.String.Interpolate  as I
+import Data.Text.Prettyprint.Doc.Extra    (Doc)
+
+
+clockWizardTclTF :: TemplateFunction
+clockWizardTclTF =
+  TemplateFunction used valid clockWizardTclTemplate
+ where
+  used = [0..4]
+  valid bbCtx
+    | [_,_,_,(nm,_,_),_,_] <- bbInputs bbCtx
+    , Just _ <- exprToString nm
+    , [(Identifier _ Nothing,Product {})] <- bbResults bbCtx
+    = True
+  valid _ = False
+
+clockWizardTclTemplate
+  :: Backend s
+  => BlackBoxContext
+  -> State s Doc
+clockWizardTclTemplate bbCtx = pure bbText
+ where
+  ((_,stripVoid -> kdIn,_):(_,stripVoid -> kdOut,_):_:(nm,_,_):_) = bbInputs bbCtx
+  Just compName = exprToString nm
+  KnownDomain _ clkInPeriod _ _ _ _ = kdIn
+  KnownDomain _ clkOutPeriod _ _ _ _ = kdOut
+  clkInFreq :: Double
+  clkInFreq  = (1.0 / (fromInteger clkInPeriod  * 1.0e-12)) / 1e6
+  clkOutFreq :: Double
+  clkOutFreq = (1.0 / (fromInteger clkOutPeriod * 1.0e-12)) / 1e6
```

Comment:
You can use [periodToHz](https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Explicit-Signal.html#v:periodToHz) here. You'll have to convert it to a `Double` if needed though.

Reply:
I suggest we leave it as is for this PR and in a future PR, define a function in `Clash.Signal.Internal` and expose it through the usual channels:

```haskell
periodToHz' :: (Integral a, Fractional b) => a -> b
periodToHz' = fromRational . toRational . periodToHz . fromIntegral
```

Or actually I'd like to propose something more radical and name it `periodToHz0` and also, for Clash 1.8, rename all functions ending in a prime symbol to functions with a `0` suffix, and then provide the prime version as perpetual synonyms for that. And never introduce any more functions ending in a prime symbol in `clash-prelude`. I'm not a big fan of prime symbols, you might have noticed.

Reply:
@DigitalBrains1 What do you think of generalizing the existing function (`periodToHz`) and adding a hint to the changelog that users can use type applications (`periodToHz @Natural @(Ratio Natural)`) to get the old behavior back - in case they get type errors?

Reply:
Yes, that's fine. People will get an extra trip through `Integer` but this is safe and the function doesn't need to perform blindingly fast at all.